### ISSUE TITLE
Docker API 1.37 returning empty IdentityToken

### ISSUE THREAD

**Issue by alex131290** [moby/moby#38830]
**Description**

Docker API v1.37 "auth" endpoint return an empty `IdentityToken`

I'm calling the Docker API `auth` endpoint from the local host

`curl -X POST http://localhost:2375/v1.37/auth -H "Content-Type: application/json" -d '{"username":"AWS", "password": "'$password'", "serveraddress": "https://<account ID>.dkr.ecr.us-east-2.amazonaws.com/"}'`
I'm getting the following output back:

`{"IdentityToken":"","Status":"Login Succeeded"}`

**Steps to reproduce the issue:**
1.
2.
3.

**Describe the results you received:**
I'm getting back `{"IdentityToken":"","Status":"Login Succeeded"}`

**Describe the results you expected:**
I'm expecting to receive back an identity `IdentityToken`

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:           18.09.3
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        774a1f4
 Built:             Thu Feb 28 06:40:58 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.3
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       774a1f4
  Built:            Thu Feb 28 05:59:55 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 2
Server Version: 18.09.3
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: nvidia runc
Default Runtime: nvidia
Init Binary: docker-init
containerd version: e6b3f5632f50dbc4e9cb6288d911bf4f5e95b18e
runc version: 12f6a991201fdb8f82579582d5e00e28fba06d0a-dirty
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-1060-aws
Operating System: Ubuntu 16.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.795GiB
Name: ip-10-220-3-78
ID: YZ6T:HXM4:XJNW:GUDY:XA6J:U2KX:R7CJ:TQHE:TPXY:HCNA:R4VL:M3AZ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

WARNING: API is accessible on http://0.0.0.0:2375 without encryption.
         Access to the remote API is equivalent to root access on the host. Refer
         to the 'Docker daemon attack surface' section in the documentation for
         more information: https://docs.docker.com/engine/security/security/#docker-daemon-attack-surface
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
I'm running on an AWS EC2 instance which has full access to ECR repos through an instance profile

**Comment by thaJeztah** [comment]
This is expected (depending on the registry that is authenticated to) see https://github.com/moby/moby/issues/33429#issuecomment-305254405 and https://github.com/moby/moby/issues/29257#issuecomment-266121974

But could perhaps the documentation around this could use some improvement to explain

**Comment by alex131290** [comment]
@thaJeztah Is it possible to communicate with ECR repositories through the API? 

Can you give an example for a build image command using authentication (using CURL for example)?

**Comment by thaJeztah** [comment]
Yes, you should be able to communicate with ECR repositories (using their `ecr` cli to authenticate; https://docs.aws.amazon.com/cli/latest/reference/ecr/get-login.html)

The `/auth` call is only to verify if the login information is incorrect; during `docker pull` / `docker push`, docker uses those credentials to create short-lived tokens.

You can have a look at this script; https://github.com/moby/moby/blob/master/contrib/download-frozen-image-v2.sh, which performs image pull without using Docker.

**Comment by alex131290** [comment]
Using `ecr` I can get the credentials In order to pass it to the `Docker login` command, but I still need to perform the build and push using the API as I'm communicating with a remote host (`ec2`).

According to the Docker API docs https://docs.docker.com/engine/api/v1.37/#operation/ImageBuild there's an option to pass the `X-Registry-Config` header , which includes the username , password and the registry , I've tried using it in order to build an image which is dependent on another image (in the `Doclerfile` - `` from my registry (the ECR repository)

**Comment by thaJeztah** [comment]
I'll do an `docker pull`, which is less complicated to show (without having to deal with uploading the Dockerfile etc), but the concept is the same:


Base64 encode the credentials (note that it should be base64-url encoded, but just to illustrate):

```bash
registryAuth=$(echo -n '{"username":"MYUSERNAME", "password": "MYPASSWORD", "serveraddress": "https://index.docker.io/v1/"}' | base64 -w0)
```

And pass those credentials in the `X-Registry-Auth` header.

This example pulls the `myname/myimage:latest` image:

```bash
curl \
  -X POST \
  --unix-socket /var/run/docker.sock \
  -H "Content-Type: application/json" \
  -H "X-Registry-Auth: $registryAuth" \
  "http://localhost/v1.37/images/create?fromImage=myname/myimage&tag=latest"
```

**Comment by alex131290** [comment]
I've tried something similar 
`password=$(aws ecr get-login --no-include-email --output text --region us-east-2 | awk '{printf $6}')`
`registryAuth=$(echo -n '{"username": "AWS", password: "'$password'", "serveraddress": "https://015895086574.dkr.ecr.us-east-2.amazonaws.com/v2/"}'| base64 -w0)`
`url -v -X POST --unix-socket /var/run/docker.sock -H "Content-Type: application/json" -H "X-Registry-Auth: $registryAuth" -H "Content-Type:application/tar" --data-binary '@Dockerfile.tar.gz' http://localhost:2375/v1.37/build`

I'm getting the same error 
`* upload completely sent off: 188 out of 188 bytes
< HTTP/1.1 200 OK
< Api-Version: 1.39
< Content-Type: application/json
< Docker-Experimental: false
< Ostype: linux
< Server: Docker/18.09.3 (linux)
< Date: Thu, 07 Mar 2019 16:02:27 GMT
< Transfer-Encoding: chunked
<
{"stream":"Step 1/1 : FROM 015895086574.dkr.ecr.us-east-2.amazonaws.com/slicer_base:latest"}
{"stream":"\n"}
{"errorDetail":{"message":"Get https://015895086574.dkr.ecr.us-east-2.amazonaws.com/v2/slicer_base/manifests/latest: no basic auth credentials"},"error":"Get https://015895086574.dkr.ecr.us-east-2.amazonaws.com/v2/slicer_base/manifests/latest: no basic auth credentials"}
* Connection #0 to host localhost left intact`
Do i need to pass the `aws ecr get-login` or the `aws ecr get-authorization-token` to the `password ` of the `registryAuth` ?

**Comment by thaJeztah** [comment]
> Do i need to pass the `aws ecr get-login` or the `aws ecr get-authorization-token` to the `password` of the `registryAuth` ?

Good question, I don't know

**Comment by alex131290** [comment]
@thaJeztah If the AWS ECR registry has been implemented only for version 2 and not 1, what would be the right command and authentication string (header) which i should pass to the docker API request ?

**Comment by alex131290** [comment]
@thaJeztah there was an issue with your example.
The registry Auth header key should be the key like the example below.
Now , it's working.
`registryAuth=$(echo -n '{ "my_account_id.dkr.ecr.us-east-2.amazonaws.com": { "username": "AWS", "password": "'$password'"}}'| base64 -w0)`
